doc/Language/haskell-to-p6.pod6

```diff
@@ -0,0 +1,420 @@
+
+TODO: include a better example for andthen that makes sense. Maybe using promise objects?
+    
+So in practice, Perl6 does not have the concept of a null type, but rather of defined or undefined types.
+
+=head2 Data Definitions
+
+Perl6 is fundamentally an Object Oriented language. However, it also gives you the freedom
+to write in virtually any paradigm you wish. If you only want to pure functions that take an
+object and return a new object, you can certainly do so.
+
+Here is a Haskell code example:
+
+    data Point = Point x y
+    
+    moveUp :: Point -> Point
+    moveUp (Point x y) = Point x (y + 1)
+    
+And an equivalent Perl6 example:
+
+    class Point { has $.x; has $.y; }
+    
+    sub move-up(Point $p --> Point) {
+      Point.new(x => $p.x, y => $p.y + 1)
+    }
+
+The code I illustrated above is an example of a Product Type. If instead you'd like to
+write a Sum Type, there is not an exact equivalent in Perl6. The closest thing would be
+an Enum.
+
+    data Animal = Dog | Cat | Bird | Horse
+    
+    testAnimal :: Animal -> String
+    testAnimal Dog   = "Woof"
+    testAnimal Horse = "Ney"
+    
+A Perl6 Enum does not fit the same exact use cases, but it can be used in putting
+constraints on types.
+
+    enum Animal < Dog Cat Bird Horse >;
+
+    proto sub test-animal( Animal        ) {*}
+    multi sub test-animal( Dog           ) { 'Woof' }
+    multi sub test-animal( Animal::Horse ) { 'Ney'  }   # more explicit
+
+    say test-animal Animal::Dog; # more explicit
+    say test-animal Horse;
+
+=head2 Type Aliases and Subsets
+
+In Haskell, you can alias an existing type to simply increase clarity of intent and re-use
+existing types.
+
+    type Name = String
+    
+    fullName :: Name -> Name -> Name
+    fullName first last = first ++ last
+    
+The equivalent in Perl6 is the following.
+
+    my constant Name = Str;
+
+    sub full-name ( Name \first, Name \last --> Name ) { first ~ last }
+    
+It should be noted that in Perl6, one can also create a subset of an existing type.
+
+    subset Name of String where *.chars < 20;
+    
+    sub full-name(Name $first, Name $last) {
+      $first ~ $last
+    }
+    
+    full-name("12345678901234567890111", "Smith") # This does not compile, as the first parameter
+                                                  # doesn't fit the Name type 
+
+=head2 Typeclasses
+
+TODO 
+
+explain how Perl6 roles compare to Haskell typeclasses
+
+=head1 Functions
+
+=head2 Definitions and Signatures
+
+=Pattern Matching
+
+Haskell makes heavy use of pattern matching in function definitions.
+
+    greeting :: String -> String
+    greeting  ""   = "Hello, World!"
+    greeting "bub" = "Hey bub."
+    greeting  name = "Hello, " ++ name ++ "!"
+    
+Perl6 does this as well! You just use the `multi` keyword to signify that it is a multiple dispatch
+function.
+
+    proto greeting ( Str   --> Str ) {*}
+    multi greeting ( ""    --> "Hello, World!" ) {}
```

Comment:
I'm fine leaving this as-is overall I suppose, the feedback that it tickled my senses is enough.

`--> True` is a fun one! It makes me feel conflicted :) . I guess in a case like that my wish would be for `multi is-thing(Thing $) { True }` to be the equivalent, so that the value and the type are statically known/inferred by the compiler.

Anyway -- is fine, I don't feel super strong about it, and it definitely _works_ the way you have it.